### DISCUSSION TITLE
Add a custom Api

### DISCUSSION THREAD

**Discussion by amitBhardwaj12** [apache/superset#18016] [Q&A / Help]
Hi, I want to add my own API that will parse the query and will return the tables list within query.
Endpoint  I am using :  '/superset/parse_query_tables/'    
POST Method 

I have added this code on backend side: 

    @has_access_api
    @event_logger.log_this
    @expose("/parse_query_tables/", methods=["POST"])
    def get_tables_list_from_query(self, query) ->FlaskResponse:
        query = request.form["sqlQuery"]
        return ParsedQuery(query).tables

Error :
Status Code: 401 UNAUTHORIZED
Error Response I am getting  :
{
  "message": "Access is Denied", 
  "severity": "danger"
}
 
Any help would be appreciated.

**Comment by eschutho** [comment]
@amitBhardwaj12 you'll need to either give your logged in user/role RBAC permission for `parse_query_tables` or map that endpoint to an existing permission set in superset/constants.py

**Reply by amitBhardwaj12** [reply]
Hi @eschutho 
My front-end point - '/superset/parse_query_tables/'

 In superset/constants.py 

API_GET_TABLES_LIST = "parse_query_tables"
MODEL_API_RW_METHOD_PERMISSION_MAP : {
 "parse_query_tables" : "read",
}



In superset/views/core.py

@has_access_api
@event_logger.log_this
@expose("/parse_query_tables/", methods=["POST"])
def get_tables_list_from_query(self, query) ->FlaskResponse:
    query = request.form["sqlQuery"]
    return ParsedQuery(query).tables
  I have added this

Its still not working for me.
can you please elaborate how should I do it.

This function : get_tables_list_from_query  ()
It will return list of all tables after parsing the query .
I want this table list on front end side.
Please help me out how my API can work.

**Reply by Gokuyamatos12** [reply]
Were you able to figure it out? I'm also trying to create a custom API endpoint in superset. @amitBhardwaj12
